Spatial search functions support multi-valued fields in compute engine

craigtaverner opened the pull request:
In #110830 it was noted that ES|QL produces different results during spatial search for multi-values when we pushdown to lucene versus when we do not. The lucene query understands spatial multi-values, while the compute engine does not.

Normally ES|QL predicates do not support multi-values because it is ambiguous whether the predicate means `ANY` or `ALL` on the multi-value. For example, if the field `age=[10,30]` and we have the query `FROM index | WHERE age > 20`, it should pass for `ANY` and fail for `ALL`. But in order to answer that, the predicate needs knowledge of this choice between `ANY` and `ALL`.

It turns out that this choice is already made for the spatial predicates. They contain knowledge of how to deal with multi-value fields (the choice of `ANY`, `ALL` or some other logic). This is because the spatial types understand the concept of a collection, and already deal with collections, so multi-valued types can be handled naturally. The lucene queries used for lucene-pushdown already support multi-values for spatial predicates like `ST_INTERSECTS`. However, the compute engine does not. The code generated from the `Evaluator` methods explicitly filters out multi-valued fields.

While this PR originally added generic support for multi-value fields using an extension to the `@Evaluator` annotation to explicitly define a combiner (Any vs All combiners), that solution was only sufficient for `ST_INTERSECTS` and `ST_DISJOINT`. It turns out that if the right-hand-side is also compound, then `ST_CONTAINS` and `ST_WITHIN` need more complex support to consider that each of the components of the other parameter needs to be contained or within at least one of the components of the first parameter. For this scenario, it is much simpler to just combined all multi-values into a single parameter before evaluation. For this reason the `@Evaluator` extension was removed in https://github.com/elastic/elasticsearch/pull/112063/commits/1792442e7338a628b2634b5132815158d6ccb692.

Instead the solution chosen here was to pass in the parameters as the original Blocks, and rely on the spatial functions to decide how to deal with them (by merging them into single compound geometries before performing the spatial relationship check). The total code complexity is lower. However, if there is interest in supporting the more general Any/All approach to other predicate functions, then reverting the above mentioned commit is one place to consider starting.

Fixes #110830
Fixed #112102
Fixes #112505

craigtaverner committed:
Using enum to control mv-predicate combinations with ANY or ALL

elasticsearchmachine commented:
Hi @craigtaverner, I've created a changelog YAML for you.

craigtaverner committed:
Update docs/changelog/112063.yaml

craigtaverner committed:
Fix changelog

craigtaverner committed:
Refactored to use generic MvCombiner for more flexibility

This opens the door to combiners which work with more types than just boolean.

craigtaverner committed:
Spotless, and disabled failing cartesian-point tests

Reported the failing cases at https://github.com/elastic/elasticsearch/issues/112102

craigtaverner committed:
Fix changelog with better summary

craigtaverner committed:
Fix changelog with better summary and highlight text

craigtaverner committed:
More spotless checks

craigtaverner committed:
Remove low-value comment edit

elasticsearchmachine commented:
Pinging @elastic/es-analytical-engine (Team:Analytics)

nik9000 commented:
> `ST_INTERSECTS(field, polygon)` - `true` if `ANY` value can intersect the polygon

What if the RHS is many polygons? Is that true if any field intersects any of the polygons?

nik9000 commented on x-pack/plugin/esql/compute/gen/src/main/java/org/elasticsearch/compute/gen/EvaluatorImplementer.java:
```diff
@@ -544,6 +606,7 @@ public void buildInvocation(StringBuilder pattern, List<Object> args, boolean bl
                 return;
             }
             if (blockStyle) {
+                // TODO: Support multiValued for non-BytesRefBlock types
```
I'd throw for now until you need it. That way the compiler will fail and folks can get an error message you wrote.

craigtaverner replied:
Added code path for multi-value case.

iverase commented:
> ST_INTERSECTS(field, polygon) - true if ANY value can intersect the polygon
> What if the RHS is many polygons? Is that true if any field intersects any of the polygons?

Yes, intersects is true if the geometries have at least one point in common.

nik9000 commented on x-pack/plugin/esql/compute/ann/src/main/java/org/elasticsearch/compute/ann/MvCombiner.java:
```diff
@@ -0,0 +1,32 @@
+public interface MvCombiner<T> {
+    T initial();
+
+    T combine(T previous, T value);
```
If we invoke these over the interface there's a good chance the JVM'll run the invocation as megamorphic. The particular way you are doing it now runs into a edge case in my knowledge - because there's always a single subclass for each call site so, theoretically, it could know that this site always uses that subclass. I'm not sure if it does though.

Either way, I think it'd be better to declare the member variable as having the exact type of the implementation. So they'd be `AllCombiner combiner = new AllCombiner()` in the class. That way it *knows*.

I know it's premature optimization, but that's kind of the point of these `ExpressionEvaluator`s - to be as efficient as we can make them.

Also! I think the `<T>` is trouble here - at least at runtime. Maybe these should just be locked to `bool` for now.  I get that this is flexible so you can do a `BytesRefBuilder` or a `long` or whatever for combination, but for now we only need `bool` and I'm pretty sure the boxing/unboxing business gives optimizers trouble.

craigtaverner replied:
This change has been implemented as a side effect of other changes I was making. We do now declare the evaluators field as the concrete type, not the interface, so hopefully these concerns are alleviated.

There might be some boxing/unboxing left, but hopefully that is a minor issue. I have taken care that we should only ever see that for multi-value cases, which right now is only for spatial search functions, so if there is any impact, it has a very narrow blast radius.

nik9000 commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/spatial/AnyCombiner.java:
```diff
@@ -0,0 +1,25 @@
+/**
+ * This multivalue combiner supports combining boolean types, where any true value will result in a true result.
+ */
+public class AnyCombiner implements MvCombiner<Boolean> {
```
I'd stick these next to `MvCombiner` over in the `compute` engine.

craigtaverner replied:
Each attempt caused dependency issues, so I've left it where it is for now.

nik9000 reviewed (approved):
I think the compute engine side is good. I left a note around performance. I know, do it right, then fast. But I'm a little paranoid about these.

It'd be really nice to have a microbenchmark too. Mostly so we can compare future work. But that can come in a follow up. Honestly, so can my point about `<T>` and declaring the actual types in the generated code.

I've not looked at any of the correctness, especially with regards to geo specifics. Just the way you are going about this in the compute engine. And that I think is a really good direction.

nik9000 commented:
> Yes, intersects is true if the geometries have at least one point in common.

:+1:

And that's what I saw when I reviewed the compute side. Thanks for confirming that's what it's supposed to do.

craigtaverner committed:
Refined MvCombiner to maintain state to deal with ST_CONTAINS

We have a special case in ST_CONTAINS in that lucenes triangle-tree implementation causes a situation where we need to reject contains results when there are other geometries that do not contain, but do intersect. This does not make sense from a pure geospatial perspective, but is a necessary consequence of the triangle-tree.

craigtaverner committed:
Merge remote-tracking branch 'origin/main' into fix_mv_spatial_predicates

craigtaverner committed:
Code review fixes, and fix for long doc-values MV

craigtaverner committed:
Cleanup and fix fold() serialization

The fold was returning the intermediate ContainsResult, which cannot be serialized, instead of the correct final boolean result.

ivancea commented on x-pack/plugin/esql/compute/ann/src/main/java/org/elasticsearch/compute/ann/Evaluator.java:
Why implementing the new combiner here instead of migrating the functions to `@MvEvaluator`?
With this, `@Evaluator` now also handles multivalues, making these annotations "overlap".

Adding to it, the Combiner working with generics and boxing also feels weird (Already commented by Nik, so nothing to add here).

So my general point here is, is it worth it to do this this way?

craigtaverner replied:
I did a brief look into MvEvaluator, but it did not seem to cover the cases I wanted. Perhaps I misunderstood things, but I did notice that some MV functions use Evaluator, not MvEvaluator, so I think there are cases where this makes sense. But it would be good to get more clarity on what the exact usages of Evaluator vs. MvEvaluator are.

craigtaverner committed:
Fix to multi-contains-multi case using BitArray

Since a multi-value field should be seen as an alternative to a geometry collection, it is insufficient to consider `ANY` for multi-value contains.

There are two approaches to this:
* Pre-build a geometry collection before converting to docValuesReader
* Maintain more state so we can assert that all components are contained within at least one of the field values

In an effort to minimize the changes to the generated code, the second approach was taken, and in fact was achievable without any changes at all to generated code.

However, this approach uses BigArrays, and does not get the correct one passed in. We need to change generated code a small bit to pass that in. We'll do that in a followup commit, but only if the alternative approach of creating a combined multi-value docValueReader is deemed more complex.

craigtaverner committed:
Fix to multi-contains-multi case using GeometryCollection

This is an alternative approach to the previous one which used a BitArray to maintain state. Now we rely entirely on the internals of the DocValuesReader, and instead pre-create the GEOMETRYCOLLECTION of all the values in the multi-value field, so the triangle tree already considers the necessary combinations.

This approach moves the responsibility of iterating over the multi-value from the generated code into the non-generated code. In total the number of lines of code goes down, as fewer code paths are possible.

craigtaverner committed:
Add addition fixed issue to changelog

craigtaverner committed:
Added csv-spec tests for testing multi-valued geometries

craigtaverner committed:
More tests for multi-value literals and one fix in fold()

craigtaverner committed:
Merge remote-tracking branch 'origin/main' into fix_mv_spatial_predicates

craigtaverner committed:
Fix bug with doc values extraction for non-indexed fields for centroid

Initially this work was about adding more tests, but discovered the bug at #112505. This commit fixes hat issue and expands the tests in a few areas:
* PhysicalPlanOptimizerTests expanded to verify that physical planning now considers if the field has doc-values
* SpatialPushDownPointsTestCase simple point-in-polygon tests expanded to consider ST_CENTROID as well, so that this behaviour is tested better there

craigtaverner committed:
Note that this PR also fixes the doc-values field extract bug

This could have been fixed in a separate PR, but fixing it here was needed because the tests we wrote were failing without it.

craigtaverner committed:
Merge remote-tracking branch 'origin/main' into fix_mv_spatial_predicates

craigtaverner committed:
Multi-point test cases

craigtaverner committed:
Added capability to prevent test failing on older clusters

Also removed a test that was sensitive to multi-node cluster results ordering

craigtaverner committed:
Support BlockBuilder multivalue combining for ST_WITHIN

This is similar too, but simpler than the ST_CONTAINS solution.
In addition we added support for two fields to handle multi-values by using ST_CONTAINS surrogate with parameters swapped.

craigtaverner committed:
Merge remote-tracking branch 'origin/main' into fix_mv_spatial_predicates

craigtaverner committed:
Require capability for BWC tests

craigtaverner committed:
Added multivalue fields tests for points

craigtaverner committed:
Support multivalues for CONTAINS/WITHIN between two fields

This included taking into account that CONTAINS and WITHIN are not symmetrical in the case that the indexed geometry contains multiple intersecting polygons.

We need to document this behaviour.

craigtaverner committed:
Merge remote-tracking branch 'origin/main' into fix_mv_spatial_predicates

craigtaverner committed:
Small optimization to not create collections over single geometries

craigtaverner committed:
Simplification of iterating over multi-value BytesRef

craigtaverner committed:
Update docs/changelog/112063.yaml

elasticsearchmachine commented:
Hi @craigtaverner, I've updated the changelog YAML for you.

craigtaverner committed:
Update docs/changelog/112063.yaml

elasticsearchmachine commented:
Hi @craigtaverner, I've updated the changelog YAML for you.

craigtaverner committed:
Added back removed bug-fix link

craigtaverner committed:
Merge branch 'fix_mv_spatial_predicates' of github.com:craigtaverner/elasticsearch into fix_mv_spatial_predicates

craigtaverner committed:
Merge conflict

iverase commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/spatial/SpatialContains.java:
```diff
@@ -87,21 +98,37 @@ protected boolean geometryRelatesGeometry(BytesRef left, BytesRef right) throws
+            boolean contains = true;
+            boolean intersectsButNotContains = false;
             for (Component2D rightComponent2D : rightComponent2Ds) {
+                boolean containsComp = geometryRelatesGeometry(leftDocValueReader, rightComponent2D);
```
I don't think this is going to work. We should try to do exactly the same thing we are doing in lucene. Therefore we should create a doc value with all the geometries and then check the spatial relationship. Otherwise there can be several edge cases which are going to be a pain.

craigtaverner replied:
I have simplified this now.

craigtaverner committed:
Support point doc-values for ST_WITHIN

craigtaverner committed:
Simplify ST_CONTAINS to not consider intersecting polygons

This turns out to already be handled by combined doc-values

craigtaverner committed:
Merge remote-tracking branch 'origin/main' into fix_mv_spatial_predicates

iverase commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/spatial/SpatialRelatesFunction.java:
```diff
@@ -188,4 +194,94 @@ protected boolean pointRelatesGeometry(Point point, Component2D component2D) {
+
+    /**
+     * When dealing with ST_CONTAINS and ST_WITHIN we need to pre-combine the field geometries for multi-values in order
+     * to perform the relationship check. This means instead of relying on the generated evaluators to iterate over all
```
Why don't we use this approach for Intersects and Contains? I would rather prefer to use the same approacch in all functions and it will remove lots of complexity too.

craigtaverner replied:
All functions now use the Block.Builder approach.

iverase commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/spatial/SpatialRelatesFunction.java:
```diff
@@ -188,4 +194,94 @@ protected boolean pointRelatesGeometry(Point point, Component2D component2D) {
+     * This class works for point and shape fields, because both are extracted into the same block encoding.
+     * However, we do detect if all values in the field are actually points and create a MultiPoint instead of a GeometryCollection.
+     */
+    protected static class MultiValuesBytesRef implements MultiValuesCombiner {
```
I wonder if this abstraction is really needed. The methods in this class can be implemented as helper methods in this class?

craigtaverner replied:
It is possible, but makes the code more complex. We have 4 methods that take the interface type `MultiValuesCombiner`), and these would need to be doubled to 8 methods (for LongBlock and BytesRefBlock versions) and each would need to add additional position parameters. I think the added complexity is not worth it.

craigtaverner committed:
Last CONTAINS evaluators moved to BlockBuilder approach

craigtaverner committed:
Merge remote-tracking branch 'origin/main' into fix_mv_spatial_predicates

craigtaverner committed:
Revert usage of MyCombiner in spatial predicates

Since ST_CONTAINS and ST_WITHIN could not use the ANY/ALL logic and needed to first collect all values into a single geometry before applying the predicate, we decided to move ST_INTERSECTS and ST_DISJOINT to this same approach so all spatial predicates have the same level of complexity and are easier to maintain.

craigtaverner committed:
Revert ability to perform ANY/ALL predicate evaluations

This was only being used by the spatial predicates, and since they have reverted to doing this logic internally, we remove this capability from the code-base. If we wish to implement ANY/ALL logic in any other predicates, this could be brought back by reverting this commit.

craigtaverner committed:
Simplify code paths for evaluators

Now that all evaluators use the Block.Builder approach we can move all the common code down to the SpatialRelations class.

This means that all static evaluator methods now contain only a single line of code, and all of them are identical between all four spatial functions, making comparison and maintenance much easier.

craigtaverner committed:
Cleanup code for easier review

craigtaverner committed:
Fixed bug with empty multivalue params and doc-values

This was failing a test in ENRICH

craigtaverner committed:
After renaming the evaluator parameters we need to update the unit tests

elasticsearchmachine commented:
## 💚 Backport successful
| Status | Branch | Result |
|:------:|:------:|:------:|
| ✅ |  [8.x](https://github.com/elastic/elasticsearch/pull/112882)  | [<img src="https://img.shields.io/github/pulls/detail/state/elastic/elasticsearch/112882">](https://github.com/elastic/elasticsearch/pull/112882) |
